### ISSUE TITLE
[SEC] Unbounded set can lead to crash nbc nodes

### ISSUE THREAD

**Issue by montyly** [status-im/nimbus-eth2#1581]
Severity: Medium
Difficulty: Medium

## Description
The code iterates over the nodes using the set operation. The operation fails if the set contains more than 65535 elements. As a result an attacker can crash a nbc node by making it iterates over a large range of value

The codebase creates frequently sets, such as `{0..self.nodes.len-1}`:
- https://github.com/status-im/nim-beacon-chain/blob/3433c77c35f3025729ac9204ab6b91ea136dce9e/beacon_chain/fork_choice/proto_array.nim#L242
- https://github.com/status-im/nim-beacon-chain/blob/3433c77c35f3025729ac9204ab6b91ea136dce9e/beacon_chain/fork_choice/proto_array.nim#L253
- https://github.com/status-im/nim-beacon-chain/blob/3433c77c35f3025729ac9204ab6b91ea136dce9e/beacon_chain/fork_choice/proto_array.nim#L308

In nim, `{a...b}` will trigger an exception if the set has more than 65535 value. 

For example the following code triggers `Error: unhandled exception: value out of range: 9999999999 notin 0 .. 65535 [RangeError]`

```nim
var 
  a: int64
  start: int64
  end_idx: int64

a = 1 
start = 2 
end_idx = 9999999999
if a notin {start..end_idx}:
  echo "not in"
echo "in"  
```

As a result, the nbc nodes will crash if they contain more than 65536 blocks. 

We classified the severity as medium as finalized blocks are pruned, reducing the likelihood to store so many blocks

## Exploit scenario
- The finalization of blocks is delayed due to an eth 2.0 issue. Many forks happen. All the nbc nodes store more than 65536 nodes and crash
- Eve finds a way to send hundred of thousands of blocks to the network. Nbc nodes store more than 65536 and all crash.

## Recommendation
Short term, replace the operations on set with direct comparison. 

Long term, report this issue to the nim team. Consider to emit a warning for set constructions that might contains more than 65536 values.

**Comment by arnetheduck** [comment]
fixed in #1535

**Comment by montyly** [comment]
Fixed in https://github.com/status-im/nimbus-eth2/pull/1535
